[CI] [Flaky test] distributed/test_shm_broadcast.py is flaky

### Anything you want to discuss about vllm.

Distributed comm ops test failed with below stacktrace. [Buildkite](https://buildkite.com/vllm/ci-aws/builds/2539#01904f5d-0bfd-4a7a-96b2-cdc7f8b60b09)

```
[2024-06-25T12:58:33Z] distributed/test_shm_broadcast.py:72:
--
  | [2024-06-25T12:58:33Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  | [2024-06-25T12:58:33Z]
  | [2024-06-25T12:58:33Z] fn = <function worker_fn_wrapper.<locals>.wrapped_fn at 0x7f8cc92afa30>
  | [2024-06-25T12:58:33Z] world_size = 4
  | [2024-06-25T12:58:33Z]
  | [2024-06-25T12:58:33Z]     def distributed_run(fn, world_size):
  | [2024-06-25T12:58:33Z]         number_of_processes = world_size
  | [2024-06-25T12:58:33Z]         processes = []
  | [2024-06-25T12:58:33Z]         for i in range(number_of_processes):
  | [2024-06-25T12:58:33Z]             env = {}
  | [2024-06-25T12:58:33Z]             env['RANK'] = str(i)
  | [2024-06-25T12:58:33Z]             env['LOCAL_RANK'] = str(i)
  | [2024-06-25T12:58:33Z]             env['WORLD_SIZE'] = str(number_of_processes)
  | [2024-06-25T12:58:33Z]             env['LOCAL_WORLD_SIZE'] = str(number_of_processes)
  | [2024-06-25T12:58:33Z]             env['MASTER_ADDR'] = 'localhost'
  | [2024-06-25T12:58:33Z]             env['MASTER_PORT'] = '12345'
  | [2024-06-25T12:58:33Z]             p = multiprocessing.Process(target=fn, args=(env, ))
  | [2024-06-25T12:58:33Z]             processes.append(p)
  | [2024-06-25T12:58:33Z]             p.start()
  | [2024-06-25T12:58:33Z]
  | [2024-06-25T12:58:33Z]         for p in processes:
  | [2024-06-25T12:58:33Z]             p.join()
  | [2024-06-25T12:58:33Z]
  | [2024-06-25T12:58:33Z]         for p in processes:
  | [2024-06-25T12:58:33Z] >           assert p.exitcode == 0
  | [2024-06-25T12:58:33Z] E           AssertionError: assert 1 == 0
  | [2024-06-25T12:58:33Z] E            +  where 1 = <Process name='Process-1' pid=15885 parent=7 stopped exitcode=1>.exitcode
```

--- a/tests/distributed/test_shm_broadcast.py
+++ b/tests/distributed/test_shm_broadcast.py
@@ -1,12 +1,22 @@
 import multiprocessing
 import random
 import time
+from typing import List
 
+import numpy as np
 import torch.distributed as dist
 
 from vllm.distributed.device_communicators.shm_broadcast import (
     ShmRingBuffer, ShmRingBufferIO)
 from vllm.utils import update_environment_variables
+
+
+def get_arrays(n: int, seed: int = 0) -> List[np.ndarray]:
+    np.random.seed(seed)
+    sizes = np.random.randint(1, 10_000, n)
+    # on average, each array will have 5k elements
+    # with int64, each array will have 40kb
+    return [np.random.randint(1, 100, i) for i in sizes]
 
 
 def distributed_run(fn, world_size):
@@ -47,24 +57,31 @@ def worker_fn_wrapper(fn):
 def worker_fn():
     writer_rank = 2
     broadcaster = ShmRingBufferIO.create_from_process_group(
-        dist.group.WORLD, 1024, 2, writer_rank)
+        dist.group.WORLD, 1024 * 1024, 2, writer_rank)
     if dist.get_rank() == writer_rank:
-        time.sleep(random.random())
-        broadcaster.broadcast_object(0)
-        time.sleep(random.random())
-        broadcaster.broadcast_object({})
-        time.sleep(random.random())
-        broadcaster.broadcast_object([])
+        seed = random.randint(0, 1000)
+        dist.broadcast_object_list([seed], writer_rank)
     else:
-        time.sleep(random.random())
-        a = broadcaster.broadcast_object(None)
-        time.sleep(random.random())
-        b = broadcaster.broadcast_object(None)
-        time.sleep(random.random())
-        c = broadcaster.broadcast_object(None)
-        assert a == 0
-        assert b == {}
-        assert c == []
+        recv = [None]
+        dist.broadcast_object_list(recv, writer_rank)
+        seed = recv[0]  # type: ignore
+    dist.barrier()
+    # in case we find a race condition
+    # print the seed so that we can reproduce the error
+    print(f"Rank {dist.get_rank()} got seed {seed}")
+    # test broadcasting with about 400MB of data
+    N = 10_000
+    if dist.get_rank() == writer_rank:
+        arrs = get_arrays(N, seed)
+        for x in arrs:
+            broadcaster.broadcast_object(x)
+            time.sleep(random.random() / 1000)
+    else:
+        arrs = get_arrays(N, seed)
+        for x in arrs:
+            y = broadcaster.broadcast_object(None)
+            assert np.array_equal(x, y)
+            time.sleep(random.random() / 1000)
     dist.barrier()
 
 

--- a/vllm/distributed/device_communicators/shm_broadcast.py
+++ b/vllm/distributed/device_communicators/shm_broadcast.py
@@ -157,7 +157,10 @@ class ShmRingBufferIO:
                     # if this block is not ready to write,
                     # we need to wait until it is read by all readers
 
-                    # wait for a while (0.1 us)
+                    # wait for a while
+                    # if we sleep for too short, it will consume too much CPU
+                    # if we sleep for too long, it will slow down the writer
+                    # 0.1 us is a good balance
                     time.sleep(1e-7)
 
                     # if we wait for a long time, we should warn the user
@@ -211,7 +214,10 @@ class ShmRingBufferIO:
                     # if this block is not ready,
                     # we need to wait until it is written
 
-                    # wait for a while (0.1 us)
+                    # wait for a while
+                    # if we sleep for too short, it will consume too much CPU
+                    # if we sleep for too long, it will slow down the reader
+                    # 0.1 us is a good balance
                     time.sleep(1e-7)
 
                     # if we wait for a long time, we should warn the user

--- a/vllm/distributed/device_communicators/shm_broadcast.py
+++ b/vllm/distributed/device_communicators/shm_broadcast.py
@@ -14,6 +14,12 @@ from vllm.logger import init_logger
 
 VLLM_RINGBUFFER_WARNING_INTERVAL = envs.VLLM_RINGBUFFER_WARNING_INTERVAL
 
+# time to wait if the queue is full or empty
+# if we sleep for too short, it will consume too much CPU
+# if we sleep for too long, it will slow down the writer/reader
+# 0.1 us is a good balance
+RINGBUFFER_SLEEP_INTERVAL = 1e-7
+
 logger = init_logger(__name__)
 
 
@@ -158,10 +164,7 @@ class ShmRingBufferIO:
                     # we need to wait until it is read by all readers
 
                     # wait for a while
-                    # if we sleep for too short, it will consume too much CPU
-                    # if we sleep for too long, it will slow down the writer
-                    # 0.1 us is a good balance
-                    time.sleep(1e-7)
+                    time.sleep(RINGBUFFER_SLEEP_INTERVAL)
 
                     # if we wait for a long time, we should warn the user
                     if time.monotonic(
@@ -215,10 +218,7 @@ class ShmRingBufferIO:
                     # we need to wait until it is written
 
                     # wait for a while
-                    # if we sleep for too short, it will consume too much CPU
-                    # if we sleep for too long, it will slow down the reader
-                    # 0.1 us is a good balance
-                    time.sleep(1e-7)
+                    time.sleep(RINGBUFFER_SLEEP_INTERVAL)
 
                     # if we wait for a long time, we should warn the user
                     if time.monotonic(

--- a/vllm/distributed/device_communicators/shm_broadcast.py
+++ b/vllm/distributed/device_communicators/shm_broadcast.py
@@ -145,7 +145,7 @@ class ShmRingBufferIO:
     @contextmanager
     def acquire_write(self):
         assert self._is_writer, "Only writers can acquire write"
-        start_time = time.time()
+        start_time = time.monotonic()
         n_warning = 1
         while True:
             with self.buffer.get_metadata(self.current_idx) as metadata_buffer:
@@ -161,7 +161,7 @@ class ShmRingBufferIO:
                     time.sleep(1e-7)
 
                     # if we wait for a long time, we should warn the user
-                    if time.time(
+                    if time.monotonic(
                     ) - start_time > VLLM_RINGBUFFER_WARNING_INTERVAL * n_warning:  # noqa
                         logger.warning(
                             "No available block found in %s second. ",
@@ -196,7 +196,7 @@ class ShmRingBufferIO:
     @contextmanager
     def acquire_read(self):
         assert self._is_reader, "Only readers can acquire read"
-        start_time = time.time()
+        start_time = time.monotonic()
         n_warning = 1
         while True:
             with self.buffer.get_metadata(self.current_idx) as metadata_buffer:
@@ -215,7 +215,7 @@ class ShmRingBufferIO:
                     time.sleep(1e-7)
 
                     # if we wait for a long time, we should warn the user
-                    if time.time(
+                    if time.monotonic(
                     ) - start_time > VLLM_RINGBUFFER_WARNING_INTERVAL * n_warning:  # noqa
                         logger.warning(
                             "No available block found in %s second. ",

--- a/vllm/distributed/device_communicators/shm_broadcast.py
+++ b/vllm/distributed/device_communicators/shm_broadcast.py
@@ -145,7 +145,6 @@ class ShmRingBufferIO:
     @contextmanager
     def acquire_write(self):
         assert self._is_writer, "Only writers can acquire write"
-        start_index = self.current_idx
         start_time = time.time()
         n_warning = 1
         while True:
@@ -154,19 +153,21 @@ class ShmRingBufferIO:
                 written_flag = metadata_buffer[0]
                 if written_flag and read_count != self.buffer.n_reader:
                     # this block is written and not read by all readers
-                    # try to write to the next block
-                    self.current_idx = (self.current_idx +
-                                        1) % self.buffer.max_chunks
-                    if self.current_idx == start_index:
-                        # no empty block found
-                        if time.time(
-                        ) - start_time > VLLM_RINGBUFFER_WARNING_INTERVAL * n_warning:  # noqa
-                            logger.warning(
-                                "No available block found in %s second. ",
-                                VLLM_RINGBUFFER_WARNING_INTERVAL)
-                            n_warning += 1
-                        # wait for a while (0.1 us)
-                        time.sleep(1e-7)
+                    # for writers, `self.current_idx` is the next block to write
+                    # if this block is not ready to write,
+                    # we need to wait until it is read by all readers
+
+                    # wait for a while (0.1 us)
+                    time.sleep(1e-7)
+
+                    # if we wait for a long time, we should warn the user
+                    if time.time(
+                    ) - start_time > VLLM_RINGBUFFER_WARNING_INTERVAL * n_warning:  # noqa
+                        logger.warning(
+                            "No available block found in %s second. ",
+                            VLLM_RINGBUFFER_WARNING_INTERVAL)
+                        n_warning += 1
+
                     continue
                 # found a block that is either
                 # (1) not written
@@ -188,12 +189,13 @@ class ShmRingBufferIO:
                     metadata_buffer[i] = 0
                 # mark the block as written
                 metadata_buffer[0] = 1
+                self.current_idx = (self.current_idx +
+                                    1) % self.buffer.max_chunks
                 break
 
     @contextmanager
     def acquire_read(self):
         assert self._is_reader, "Only readers can acquire read"
-        start_index = self.current_idx
         start_time = time.time()
         n_warning = 1
         while True:
@@ -204,19 +206,22 @@ class ShmRingBufferIO:
                     # this block is either
                     # (1) not written
                     # (2) already read by this reader
-                    # try to read the next block
-                    self.current_idx = (self.current_idx +
-                                        1) % self.buffer.max_chunks
-                    if self.current_idx == start_index:
-                        # no block found
-                        if time.time(
-                        ) - start_time > VLLM_RINGBUFFER_WARNING_INTERVAL * n_warning:  # noqa
-                            logger.warning(
-                                "No available block found in %s second. ",
-                                VLLM_RINGBUFFER_WARNING_INTERVAL)
-                            n_warning += 1
-                        # wait for a while (0.1 us)
-                        time.sleep(1e-7)
+
+                    # for readers, `self.current_idx` is the next block to read
+                    # if this block is not ready,
+                    # we need to wait until it is written
+
+                    # wait for a while (0.1 us)
+                    time.sleep(1e-7)
+
+                    # if we wait for a long time, we should warn the user
+                    if time.time(
+                    ) - start_time > VLLM_RINGBUFFER_WARNING_INTERVAL * n_warning:  # noqa
+                        logger.warning(
+                            "No available block found in %s second. ",
+                            VLLM_RINGBUFFER_WARNING_INTERVAL)
+                        n_warning += 1
+
                     continue
                 # found a block that is not read by this reader
                 # let caller read from the buffer
@@ -226,6 +231,8 @@ class ShmRingBufferIO:
                 # caller has read from the buffer
                 # set the read flag
                 metadata_buffer[self.reader_rank + 1] = 1
+                self.current_idx = (self.current_idx +
+                                    1) % self.buffer.max_chunks
                 break
 
     def enqueue(self, obj):

